[sig-scheduling] SchedulerPriorities [Serial] Pod should avoid nodes that have avoidPod annotation: clean remaining pods

ingvagabund opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
In some cases when running "[sig-scheduling] SchedulerPredicates [Serial] validates resource limits of pods that are allowed to run"
"scheduler-priority-avoid-pod-XXX" filler pods from "[sig-scheduling] SchedulerPriorities [Serial] Pod should avoid nodes that have avoidPod annotation" are still being
terminated in the cluster. Which can cause "validates resource limits of pods that are allowed to run" to flake since the filler
pods dissappear before "additional-pod" is scheduled. Causing the calculated filler-pods to consume less than they are supposed to.

/kind bug

**What this PR does / why we need it**:
Reduce flakiness of `[sig-scheduling] SchedulerPredicates [Serial] validates resource limits of pods that are allowed to run`


**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #94715

**Special notes for your reviewer**:

Scheduler logs around the time of flakiness:
```
I0904 13:19:21.079071       1 debugger.go:65] "Dumping node infos" node="master-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[memory=5579Mi pods=0 ephemeral-storage=0 cpu=1895m]"
I0904 13:19:21.079095       1 debugger.go:65] "Dumping node infos" node="master-1.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[cpu=1720m memory=5049Mi pods=0 ephemeral-storage=0]"
I0904 13:19:21.079108       1 debugger.go:65] "Dumping node infos" node="master-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[ephemeral-storage=0 cpu=1934m memory=5999Mi pods=0]"
I0904 13:19:21.079121       1 debugger.go:65] "Dumping node infos" node="worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[cpu=773m memory=3097Mi pods=0 ephemeral-storage=0]"
I0904 13:19:21.079133       1 debugger.go:65] "Dumping node infos" node="worker-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[memory=3282Mi pods=0 ephemeral-storage=0 cpu=572m]"
I0904 13:19:21.079145       1 debugger.go:65] "Dumping node infos" node="worker-1.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[cpu=468m memory=2931Mi pods=0 ephemeral-storage=0]"
I0904 13:19:21.083259       1 scheduler.go:598] "Successfully bound pod to node" pod="e2e-sched-priority-3802/ead2872e-4ad3-4244-bb02-9f7ff0b6a428-0" node="worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" evaluatedNodes=6 feasibleNodes=1
I0904 13:19:26.257626       1 debugger.go:65] "Dumping node infos" node="master-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[cpu=1895m memory=5579Mi pods=0 ephemeral-storage=0]"
I0904 13:19:26.257646       1 debugger.go:65] "Dumping node infos" node="master-1.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[cpu=1720m memory=5049Mi pods=0 ephemeral-storage=0]"
I0904 13:19:26.257656       1 debugger.go:65] "Dumping node infos" node="master-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[cpu=1934m memory=5999Mi pods=0 ephemeral-storage=0]"
I0904 13:19:26.257668       1 debugger.go:65] "Dumping node infos" node="worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[cpu=3650m memory=14938440Ki pods=0 ephemeral-storage=0]"
I0904 13:19:26.257677       1 debugger.go:65] "Dumping node infos" node="worker-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[cpu=572m memory=3282Mi pods=0 ephemeral-storage=0]"
I0904 13:19:26.257689       1 debugger.go:65] "Dumping node infos" node="worker-1.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[memory=2931Mi pods=0 ephemeral-storage=0 cpu=468m]"
I0904 13:19:26.263121       1 scheduler.go:598] "Successfully bound pod to node" pod="e2e-sched-priority-3802/67e791a6-f170-4aff-b2b1-e42a0a6a7c3c-0" node="worker-1.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" evaluatedNodes=6 feasibleNodes=1
I0904 13:19:31.436220       1 debugger.go:65] "Dumping node infos" node="master-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[cpu=1895m memory=5579Mi pods=0 ephemeral-storage=0]"
I0904 13:19:31.436239       1 debugger.go:65] "Dumping node infos" node="master-1.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[ephemeral-storage=0 cpu=1720m memory=5049Mi pods=0]"
I0904 13:19:31.436251       1 debugger.go:65] "Dumping node infos" node="master-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[cpu=1934m memory=5999Mi pods=0 ephemeral-storage=0]"
I0904 13:19:31.436264       1 debugger.go:65] "Dumping node infos" node="worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[ephemeral-storage=0 cpu=3650m memory=14938440Ki pods=0]"
I0904 13:19:31.436275       1 debugger.go:65] "Dumping node infos" node="worker-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[cpu=572m memory=3282Mi pods=0 ephemeral-storage=0]"
I0904 13:19:31.436288       1 debugger.go:65] "Dumping node infos" node="worker-1.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[cpu=3650m memory=14874430Ki pods=0 ephemeral-storage=0]"
I0904 13:19:31.440420       1 scheduler.go:598] "Successfully bound pod to node" pod="e2e-sched-priority-3802/ffecd964-0982-4541-ba35-bc53ef1bcd52-0" node="worker-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" evaluatedNodes=6 feasibleNodes=1
I0904 13:19:40.139882       1 debugger.go:65] "Dumping node infos" node="master-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[cpu=1895m memory=5579Mi pods=0 ephemeral-storage=0]"
I0904 13:19:40.139904       1 debugger.go:65] "Dumping node infos" node="master-1.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[cpu=1720m memory=5049Mi pods=0 ephemeral-storage=0]"
I0904 13:19:40.139914       1 debugger.go:65] "Dumping node infos" node="master-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[cpu=1934m memory=5999Mi pods=0 ephemeral-storage=0]"
I0904 13:19:40.139922       1 debugger.go:65] "Dumping node infos" node="worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[pods=0 ephemeral-storage=0 cpu=3650m memory=14938440Ki]"
I0904 13:19:40.139930       1 debugger.go:65] "Dumping node infos" node="worker-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[memory=15231428607 pods=0 ephemeral-storage=0 cpu=3650m]"
I0904 13:19:40.139942       1 debugger.go:65] "Dumping node infos" node="worker-1.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[pods=0 ephemeral-storage=0 cpu=3650m memory=14874430Ki]"
I0904 13:19:40.145351       1 scheduler.go:598] "Successfully bound pod to node" pod="e2e-sched-priority-3802/scheduler-priority-avoid-pod-j72gx" node="worker-1.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" evaluatedNodes=6 feasibleNodes=3
I0904 13:19:40.146963       1 debugger.go:65] "Dumping node infos" node="master-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[pods=0 ephemeral-storage=0 cpu=1895m memory=5579Mi]"
I0904 13:19:40.147004       1 debugger.go:65] "Dumping node infos" node="master-1.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[pods=0 ephemeral-storage=0 cpu=1720m memory=5049Mi]"
I0904 13:19:40.147026       1 debugger.go:65] "Dumping node infos" node="master-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[cpu=1934m memory=5999Mi pods=0 ephemeral-storage=0]"
I0904 13:19:40.147046       1 debugger.go:65] "Dumping node infos" node="worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[cpu=3650m memory=14938440Ki pods=0 ephemeral-storage=0]"
I0904 13:19:40.147068       1 debugger.go:65] "Dumping node infos" node="worker-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[memory=15231428607 pods=0 ephemeral-storage=0 cpu=3650m]"
I0904 13:19:40.147092       1 debugger.go:65] "Dumping node infos" node="worker-1.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[memory=14976830Ki pods=0 ephemeral-storage=0 cpu=3750m]"
I0904 13:19:40.150793       1 scheduler.go:598] "Successfully bound pod to node" pod="e2e-sched-priority-3802/scheduler-priority-avoid-pod-j6kgx" node="worker-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" evaluatedNodes=6 feasibleNodes=3
I0904 13:20:00.462039       1 debugger.go:65] "Dumping node infos" node="master-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[cpu=1895m memory=5579Mi pods=0 ephemeral-storage=0]"
I0904 13:20:00.462066       1 debugger.go:65] "Dumping node infos" node="master-1.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[cpu=1720m memory=5049Mi pods=0 ephemeral-storage=0]"
I0904 13:20:00.462079       1 debugger.go:65] "Dumping node infos" node="master-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[cpu=1934m memory=5999Mi pods=0 ephemeral-storage=0]"
I0904 13:20:00.462091       1 debugger.go:65] "Dumping node infos" node="worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[pods=0 ephemeral-storage=0 cpu=3650m memory=14938440Ki]"
I0904 13:20:00.462103       1 debugger.go:65] "Dumping node infos" node="worker-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[cpu=3650m memory=15231428607 pods=0 ephemeral-storage=0]"
I0904 13:20:00.462122       1 debugger.go:65] "Dumping node infos" node="worker-1.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[ephemeral-storage=0 cpu=3650m memory=14874430Ki pods=0]"
I0904 13:20:00.466956       1 scheduler.go:598] "Successfully bound pod to node" pod="e2e-sched-pred-4521/filler-pod-5be568c0-1552-4f2e-8ff2-869c5cd610f7" node="worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" evaluatedNodes=6 feasibleNodes=1
I0904 13:20:00.541768       1 debugger.go:65] "Dumping node infos" node="master-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[pods=0 ephemeral-storage=0 cpu=1895m memory=5579Mi]"
I0904 13:20:00.541793       1 debugger.go:65] "Dumping node infos" node="master-1.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[cpu=1720m memory=5049Mi pods=0 ephemeral-storage=0]"
I0904 13:20:00.541806       1 debugger.go:65] "Dumping node infos" node="master-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[cpu=1934m memory=5999Mi pods=0 ephemeral-storage=0]"
I0904 13:20:00.541818       1 debugger.go:65] "Dumping node infos" node="worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[cpu=6345m memory=14938440Ki pods=0 ephemeral-storage=0]"
I0904 13:20:00.541833       1 debugger.go:65] "Dumping node infos" node="worker-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[cpu=3650m memory=15231428607 pods=0 ephemeral-storage=0]"
I0904 13:20:00.541847       1 debugger.go:65] "Dumping node infos" node="worker-1.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[cpu=3650m memory=14874430Ki pods=0 ephemeral-storage=0]"
I0904 13:20:00.545296       1 scheduler.go:598] "Successfully bound pod to node" pod="e2e-sched-pred-4521/filler-pod-8e8309c7-b858-481c-8052-246ba24580d0" node="worker-1.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" evaluatedNodes=6 feasibleNodes=1
I0904 13:20:00.622148       1 debugger.go:65] "Dumping node infos" node="master-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[memory=5579Mi pods=0 ephemeral-storage=0 cpu=1895m]"
I0904 13:20:00.622175       1 debugger.go:65] "Dumping node infos" node="master-1.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[memory=5049Mi pods=0 ephemeral-storage=0 cpu=1720m]"
I0904 13:20:00.622188       1 debugger.go:65] "Dumping node infos" node="master-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[ephemeral-storage=0 cpu=1934m memory=5999Mi pods=0]"
I0904 13:20:00.622199       1 debugger.go:65] "Dumping node infos" node="worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[cpu=6345m memory=14938440Ki pods=0 ephemeral-storage=0]"
I0904 13:20:00.622213       1 debugger.go:65] "Dumping node infos" node="worker-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[cpu=3650m memory=15231428607 pods=0 ephemeral-storage=0]"
I0904 13:20:00.622229       1 debugger.go:65] "Dumping node infos" node="worker-1.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[pods=0 ephemeral-storage=0 cpu=6345m memory=14874430Ki]"
I0904 13:20:00.625867       1 scheduler.go:598] "Successfully bound pod to node" pod="e2e-sched-pred-4521/filler-pod-e98507a6-2221-4312-9f1b-34b7d1124b91" node="worker-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" evaluatedNodes=6 feasibleNodes=1
I0904 13:20:05.309985       1 debugger.go:65] "Dumping node infos" node="master-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[cpu=1895m memory=5579Mi pods=0 ephemeral-storage=0]"
I0904 13:20:05.310011       1 debugger.go:65] "Dumping node infos" node="master-1.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[cpu=1720m memory=5049Mi pods=0 ephemeral-storage=0]"
I0904 13:20:05.310021       1 debugger.go:65] "Dumping node infos" node="master-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[ephemeral-storage=0 cpu=1934m memory=5999Mi pods=0]"
I0904 13:20:05.310029       1 debugger.go:65] "Dumping node infos" node="worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[cpu=3468m memory=3097Mi pods=0 ephemeral-storage=0]"
I0904 13:20:05.310037       1 debugger.go:65] "Dumping node infos" node="worker-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[ephemeral-storage=0 cpu=3267m memory=3282Mi pods=0]"
I0904 13:20:05.310044       1 debugger.go:65] "Dumping node infos" node="worker-1.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" requested="[pods=0 ephemeral-storage=0 cpu=6345m memory=14874430Ki]"
I0904 13:20:05.314450       1 scheduler.go:598] "Successfully bound pod to node" pod="e2e-sched-pred-4521/additional-pod" node="worker-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com" evaluatedNodes=6 feasibleNodes=2
```

Notice the change of cpu on `worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com` right before `e2e-sched-pred-4521/additional-pod` is scheduled. Requested cpu of all pods on the node dropped from `cpu=6345m` to `cpu=3468m` as `ead2872e-4ad3-4244-bb02-9f7ff0b6a428-0` pod finally disappeared.

Before that the total requested cpu was computed through:
```
Sep  4 13:20:00.412: INFO: Pod 67e791a6-f170-4aff-b2b1-e42a0a6a7c3c-0 requesting resource cpu=3182m on Node worker-1.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod ead2872e-4ad3-4244-bb02-9f7ff0b6a428-0 requesting resource cpu=2877m on Node worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod ffecd964-0982-4541-ba35-bc53ef1bcd52-0 requesting resource cpu=3078m on Node worker-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod tuned-cmmn2 requesting resource cpu=10m on Node worker-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod tuned-tljds requesting resource cpu=10m on Node worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod tuned-v48gt requesting resource cpu=10m on Node worker-1.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod dns-default-rc6ht requesting resource cpu=65m on Node worker-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod dns-default-tbqpg requesting resource cpu=65m on Node worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod dns-default-x9v97 requesting resource cpu=65m on Node worker-1.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod image-registry-7cf7d78fd6-4mxr9 requesting resource cpu=100m on Node worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod node-ca-72247 requesting resource cpu=10m on Node worker-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod node-ca-8nmtz requesting resource cpu=10m on Node worker-1.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod node-ca-zr995 requesting resource cpu=10m on Node worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod router-default-556cc9b846-8zrxd requesting resource cpu=100m on Node worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod router-default-556cc9b846-x8j5j requesting resource cpu=100m on Node worker-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod migrator-865bcf9488-tkjff requesting resource cpu=100m on Node worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod machine-config-daemon-7gnl4 requesting resource cpu=40m on Node worker-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod machine-config-daemon-n895c requesting resource cpu=40m on Node worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod machine-config-daemon-wf995 requesting resource cpu=40m on Node worker-1.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod certified-operators-c4kzf requesting resource cpu=10m on Node worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod certified-operators-whjv4 requesting resource cpu=10m on Node worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod community-operators-88cwp requesting resource cpu=10m on Node worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod community-operators-qvc86 requesting resource cpu=10m on Node worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod redhat-marketplace-dxk6r requesting resource cpu=10m on Node worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod redhat-marketplace-fnsfv requesting resource cpu=10m on Node worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod redhat-operators-n9vwc requesting resource cpu=10m on Node worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod redhat-operators-xz9qd requesting resource cpu=10m on Node worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod alertmanager-main-0 requesting resource cpu=8m on Node worker-1.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod alertmanager-main-1 requesting resource cpu=8m on Node worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod alertmanager-main-2 requesting resource cpu=8m on Node worker-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod grafana-5dc687d5df-mltbk requesting resource cpu=5m on Node worker-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod kube-state-metrics-747cdddc64-frwfw requesting resource cpu=4m on Node worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod node-exporter-mjnv9 requesting resource cpu=9m on Node worker-1.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod node-exporter-rklqh requesting resource cpu=9m on Node worker-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod node-exporter-w8279 requesting resource cpu=9m on Node worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod openshift-state-metrics-59c8cd5968-5rdz7 requesting resource cpu=3m on Node worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod prometheus-adapter-6b7f847c77-pc945 requesting resource cpu=1m on Node worker-1.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod prometheus-adapter-6b7f847c77-xs5kb requesting resource cpu=1m on Node worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod prometheus-k8s-0 requesting resource cpu=76m on Node worker-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod prometheus-k8s-1 requesting resource cpu=76m on Node worker-1.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod telemeter-client-657bd8cbd6-7ll4v requesting resource cpu=3m on Node worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod thanos-querier-5c57566d79-c5n5d requesting resource cpu=9m on Node worker-1.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod thanos-querier-5c57566d79-zc2sf requesting resource cpu=9m on Node worker-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod multus-9p7zj requesting resource cpu=10m on Node worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod multus-b6z24 requesting resource cpu=10m on Node worker-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod multus-h7dtf requesting resource cpu=10m on Node worker-1.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod network-metrics-daemon-9nvjv requesting resource cpu=20m on Node worker-1.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod network-metrics-daemon-cmnw6 requesting resource cpu=20m on Node worker-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod network-metrics-daemon-k2xxs requesting resource cpu=20m on Node worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod ovs-k7n5z requesting resource cpu=100m on Node worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod ovs-shgxf requesting resource cpu=100m on Node worker-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod ovs-ssnc6 requesting resource cpu=100m on Node worker-1.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod sdn-sblm6 requesting resource cpu=110m on Node worker-2.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod sdn-wzgwj requesting resource cpu=110m on Node worker-0.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
Sep  4 13:20:00.412: INFO: Pod sdn-xw75k requesting resource cpu=110m on Node worker-1.ci-op-qfypnb1x-18a82.origin-ci-int-aws.dev.rhcloud.com
```

where clearly `67e791a6-f170-4aff-b2b1-e42a0a6a7c3c-0`, `ead2872e-4ad3-4244-bb02-9f7ff0b6a428-0` and `ffecd964-0982-4541-ba35-bc53ef1bcd52-0` are taken into account when they should not.

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

ingvagabund commented:
@alculquicondor PTAL

alculquicondor commented on test/e2e/scheduling/priorities.go:
```diff
@@ -252,6 +252,21 @@ var _ = SIGDescribe("SchedulerPriorities [Serial]", func() {
+			// terminating state, give them more time to get deleted so subsequent tests
+			// (e.g. [sig-scheduling] SchedulerPredicates [Serial] validates resource limits of pods that are allowed to run)
+			// do not process pods that are disappearing under their hands.
+			wait.PollImmediate(time.Second, 2*time.Minute, func() (bool, error) {
```
Can you look into fixing `DeleteResourceAndWaitForGC` instead? According to its comment, it should already wait for it. Are you seeing `Failed to cleanup replication controller` in the logs? If so, we should fix those.

ingvagabund replied:
Unfortunately, I have logs only from failed ci runs. `[sig-scheduling] SchedulerPredicates [Serial] validates resource limits of pods that are allowed to run` is green when `[sig-scheduling] SchedulerPredicates [Serial] validates resource limits of pods that are allowed to run` flakes. Also, it's reproducible only in ~5% of cases.

alculquicondor replied:
In that case, I'm not convinced that this patch will solve the flakiness, as we don't understand the root cause of the leaking pods.

alculquicondor replied:
Could you share a failed run?

ingvagabund replied:
> Could you share a failed run?

In case of `[sig-scheduling] SchedulerPriorities [Serial] Pod should avoid nodes that have avoidPod annotation`, it will never fail because of pods not getting properly deleted. 

In case of `[sig-scheduling] SchedulerPredicates [Serial] validates resource limits of pods that are allowed to run`, just reported https://github.com/kubernetes/kubernetes/issues/94715

ingvagabund replied:
@alculquicondor what about to fail when `e2erc.DeleteRCAndWaitForGC` returns an error instead? It's not a failure related to the test itself. Though, the test should still make sure all the resources created by it are cleaned.

alculquicondor replied:
What we need to know is why it fails and fix that, because that's the function that is intended to do the cleanup. By waiting again after that function, we are duplicating code and moving the issue somewhere else.

ingvagabund replied:
@alculquicondor I still have not been able to reproduce the case and get the error message when `e2erc.DeleteRCAndWaitForGC` is not deleting all the pods. Instead, I am updating `validates resource limits of pods that are allowed to run` to ignore all pods with `pod.DeletionTimestamp` set. Since those pods can disappear at any moment and make the calculated total resources unusable for decision making.

alculquicondor commented:
/remove-kind bug
/kind flake

alculquicondor commented on test/e2e/scheduling/predicates.go:
```diff
@@ -356,7 +356,7 @@ var _ = SIGDescribe("SchedulerPredicates [Serial]", func() {
 		for _, pod := range pods.Items {
 			_, found := nodeToAllocatableMap[pod.Spec.NodeName]
-			if found && pod.Status.Phase != v1.PodSucceeded && pod.Status.Phase != v1.PodFailed {
+			if found && pod.Status.Phase != v1.PodSucceeded && pod.Status.Phase != v1.PodFailed && pod.DeletionTimestamp == nil {
```
This might cause another flake where the last pod is scheduled if one Pod with DeletionTimestamp doesn't ever terminate.

Have you tried increasing timeouts in the cleanup function for replication controller?

ingvagabund replied:
> This might cause another flake where the last pod is scheduled if one Pod with DeletionTimestamp doesn't ever terminate.

If the pod with DeletionTimestamp set does not terminate, there will be even more resource requested. I.e. resource requested by all pods without DeletionTimestamp set + filler pods + pods with DeletionTimestamp set.

alculquicondor replied:
Sorry, the case I was thinking of is actually this:

There are a lot of Pods with DeletionTimestamp initially, and they they all terminate. Thus there is a lot of space and the last pod gets scheduled.

ingvagabund replied:
Let's use `pods` for a set of all pods (including pods with `DeletionTimestamp` set) and `podswithoutDT` for a set of all pods with `DeletionTimestamp` set to nil

```
for _, pod := range pods.Items {
	_, found := nodeToAllocatableMap[pod.Spec.NodeName]
	if found && pod.Status.Phase != v1.PodSucceeded && pod.Status.Phase != v1.PodFailed && pod.DeletionTimestamp == nil {
		framework.Logf("Pod %v requesting resource cpu=%vm on Node %v", pod.Name, getRequestedCPU(pod), pod.Spec.NodeName)
		nodeToAllocatableMap[pod.Spec.NodeName] -= getRequestedCPU(pod)
	}
}
```

After the cycle `nodeToAllocatableMap[node]` constructed from `pods` will be less or equal to the same map constructed from `podswithoutDT`. I.e. `nodeToAllocatableMap_pods[node] <= nodeToAllocatableMap_podswithoutDT[node]`.

With the next cycle where the `filler-pod-XXX` pods are constructed:
```
for nodeName, cpu := range nodeToAllocatableMap {
	requestedCPU := cpu * 7 / 10
...
```

Given `nodeToAllocatableMap_pods[node] <= nodeToAllocatableMap_podswithoutDT[node]`, also `requestedCPU` for set `pods` will be less or equal to `podswithoutDT`. I.e.

`requestedCPU_pods <= requestedCPU_podswithoutDT`.

In other words, requested cpu of a filler pod for all pods including pods with DeletionTimestamp set (on a given node) will never be more than requested cpu for all pods (on a given node) without pods with DeletionTimestamp set. So if any of the pods with DeletionTimestamp set disappears, the requested cpu (from `podswithoutDT` set) of a filler pod will compensate. 

It's actually the opposite problem in question. What if the pods with DeletionTimestamp set do not disappear before the `additional-pod` is scheduled? `additional-pod` will not still get scheduled since there's gonna be even less of cpu resource left. The worst case scenario is when there's so many pods with DeletionTimestamp set that even the filler pods will not get scheduled.

ingvagabund replied:
> There are a lot of Pods with DeletionTimestamp initially and they all terminate after this loop. Thus there is a lot of space and the last pod gets scheduled.

This is actually the case happening now

alculquicondor replied:
Thanks for detailed analysis. You are right, the scenario where `additional-pod` doesn't get scheduled is possible. It sounds like the only correct solution is to fix the cleanup of other tests :/

ingvagabund replied:
I am afraid there can be other tests not cleaning their pods properly in the future (or maybe even now). If so, this needs to be fixed on multiple places. I'd start here since this change is making the test less flaky if there's not so many pods with DeletionTimestamp set.

alculquicondor replied:
I'm not comfortable with this patch. @Huang-Wei @ahg-g to comment.

alculquicondor replied:
Even if this makes the test less flaky, that's not necessarily better, as we might be hiding legitimate bugs.

ahg-g replied:
>  The worst case scenario is when there's so many pods with DeletionTimestamp set that even the filler pods will not get scheduled.

Right, and this will most likely cause flakes in the future.

We already call `WaitForStableCluster`, perhaps we can modify it to call `WaitForPodsToBeDeleted`? If those pods don't get deleted, then there is something wrong happening and we can address that instead of working around it by ignoring pods to be deleted.

Huang-Wei replied:
> We already call `WaitForStableCluster`, perhaps we can modify it to call `WaitForPodsToBeDeleted`?

+1, instead of embedding `WaitForPodsToBeDeleted` into `WaitForStableCluster`, another option is to call `WaitForPodsToBeDeleted` on demand, immediately after `WaitForStableCluster`, for example in the case.

BTW: `WaitForPodsToBeDeleted` was used in an old test case (which is deleted now) in 1.17:

https://github.com/kubernetes/kubernetes/blob/5ec472285121eb6c451e515bc0a7201413872fa3/test/e2e/scheduling/predicates.go#L132

So I _guess_ `WaitForPodsToBeDeleted` was introduced at that time to resolve similar flakes.

Huang-Wei replied:
Another thing a bit off the topic: both `WaitForStableCluster` and `WaitForPodsToBeDeleted` are performing an infinite `for` loop, although the ginkgo test may be guarded by a timeout argument, I think it'd be good to explicitly time out the test inside the function. Anyway, it can be a follow-up.

ingvagabund replied:
I was able to reproduce the test locally. The remaining pods (`67e791a6-f170-4aff-b2b1-e42a0a6a7c3c-0`, `ead2872e-4ad3-4244-bb02-9f7ff0b6a428-0` and `ffecd964-0982-4541-ba35-bc53ef1bcd52-0`) were not deleted before the testing namespace was. Thus, leaving the pods still running even after the test ended. PR update to delete all remaining pods in the testing namespace first.

@alculquicondor @ahg-g @Huang-Wei  PTAL

alculquicondor replied:
I still have the same issue with this patch.

Why don't we fix `DeleteRCAndWaitForGC` instead?

ingvagabund replied:
The remaining pods have no owner

alculquicondor replied:
can you tell what created them then?

Again, I'd like to fix the root of the problem.

ingvagabund replied:
https://github.com/kubernetes/kubernetes/blob/64c099d67069f85f2b9b78d7d12fd1a92fcc3d75/test/e2e/scheduling/priorities.go#L484

Other option is to have `createBalancedPodForNodes` return cleanup function that will be called in `defer` block.

alculquicondor replied:
oh, I get it now. These Pods are not created for the RC.

2 things then:
- Use `DeleteCollection()`.
- Do this in a separate defer, right after the Pods are created. Otherwise if RC creation fails, these Pods don't get cleaned.

Maybe for follow up: could you check for other instances where `createBalancedPodForNodes` is called and Pods are not cleared? Is this ever intended?

ingvagabund replied:
`DeleteCollection` does not wait until all the pods are deleted. Which makes it unusable without additional waiting logic.

ingvagabund replied:
PR updated. `createBalancedPodForNodes` now returns cleanUp function which is called right away through defer.

ingvagabund committed:
e2e: Pod should avoid nodes that have avoidPod annotation: clean remaining pods

The test is not cleaning all pods it created.
Memory balancing pods are deleted once the test namespace is.
Thus, leaving the pods running or in terminating state when a new test is run.
In case the next test is "[sig-scheduling] SchedulerPredicates [Serial] validates resource limits of pods that are allowed to run",
the test can fail.

alculquicondor commented on test/e2e/scheduling/predicates.go:
```diff
@@ -887,8 +887,8 @@ func initPausePod(f *framework.Framework, conf pausePodConfig) *v1.Pod {
 			Namespace:       conf.Namespace,
-			Labels:          conf.Labels,
-			Annotations:     conf.Annotations,
+			Labels:          map[string]string{},
```
why is this change necessary?

ingvagabund replied:
`testutils.StartPods` modifies Labels field. `Labels: conf.Labels` is a shallow copy and `conf.Labels` is a [shallow copy](https://github.com/kubernetes/kubernetes/blob/64c099d67069f85f2b9b78d7d12fd1a92fcc3d75/test/e2e/scheduling/priorities.go#L464) of `balancePodLabel`. Every call to `testutils.StartPods` then modifies `balancePodLabel` and disallows `balancePodLabel` to be used for listing all pods created for memory balancing.

alculquicondor replied:
sg

alculquicondor commented on test/e2e/scheduling/priorities.go:
```diff
@@ -425,7 +430,23 @@ var _ = SIGDescribe("SchedulerPriorities [Serial]", func() {
+			framework.Logf("Failed to list remaining pods for cleanup: %v.", err)
+		} else {
+			for _, pod := range podList.Items {
+				if err := e2epod.DeletePodWithWaitByName(cs, pod.Name, pod.Namespace); err != nil {
```
can we parallelize and have a wait group?
Alternatively, you can do `DeleteCollection` and have a `wait.Poll` that does a `List` until it returns empty.

ingvagabund replied:
Done

alculquicondor reviewed (commented):
/approve
/lgtm

Thanks a lot for all the debugging and commitment to solve this flaky!
